### PR TITLE
getter and setters for ptk, ccmp, keys, pmks, capturer and aps

### PR DESCRIPTION
**Pull request by prasad24582**
Hi Matias,
Am creating this pull request to add support to getter and setters to save/captured keys and if we want to override decrypt method, we can implement our own decrypt function.

Please let me know if I miss any.
Thank you for your patience.

Thanks and Regards,
Prasad

### CODE REVIEW

File: include/tins/crypto.h [hunk 1]
```diff
@@ -231,14 +258,67 @@ namespace Crypto {
          * \return false if no decryption was performed, or the decryption 
          * failed, true otherwise.
          */
-        bool decrypt(PDU &pdu);
-    private:
-        typedef std::map<std::string, WPA2::SupplicantData> pmks_map;
-        typedef std::map<address_type, WPA2::SupplicantData> bssids_map;
-        typedef std::pair<address_type, address_type> addr_pair;
-        typedef std::map<addr_pair, WPA2::SessionKeys> keys_map;
+        virtual bool decrypt(PDU &pdu);
```
Comment (mfontanini) [inline include/tins/crypto.h]: I think we agreed on this not being virtual

File: include/tins/crypto.h [hunk 2]
```diff
@@ -66,7 +66,14 @@ namespace Crypto {
             
             SessionKeys();
             SessionKeys(const RSNHandshake &hs, const pmk_type &pmk);
+            SessionKeys(const ptk_type &rptk, const bool ccmp);
```
Comment (mfontanini) [inline include/tins/crypto.h]: Just use ```bool``` here, no need for const.

File: include/tins/crypto.h [hunk 3]
```diff
@@ -66,7 +66,14 @@ namespace Crypto {
...
             SNAP *decrypt_unicast(const Dot11Data &dot11, RawPDU &raw) const;
+            const ptk_type &get_ptk() const {
+                return ptk;
+            }
+            const bool &get_ccmp() const {
```
Comment (mfontanini) [inline include/tins/crypto.h]: This should return ```bool``` rather than ```const bool&```.

File: include/tins/crypto.h [hunk 4]
```diff
@@ -232,13 +259,66 @@ namespace Crypto {
...
+         */
+        void add_access_point(const std::string &ssid, const address_type &addr);
+        
+        /**
+         * \brief get pmks.
+         * 
+         * return pmks.
+         */
+        pmks_map &get_pmks() {
+          return pmks;
```
Comment (mfontanini) [inline include/tins/crypto.h]: This line doesn't seem to be padded to 4 spaces like the rest of the code.

File: include/tins/crypto.h [hunk 5]
```diff
@@ -232,13 +259,66 @@ namespace Crypto {
...
+        }
+        
+        /**
+         * \brief get keys map.
+         * 
+         * return keys to save captured keys or to set provious stored keys.
+         */
+         keys_map &get_keys(){
+          return keys;
+        }
```
Comment (mfontanini) [inline include/tins/crypto.h]: Spacing issue here too

File: include/tins/crypto.h [hunk 6]
```diff
@@ -232,13 +259,66 @@ namespace Crypto {
...
+        
+        /**
+         * \brief get capturer.
+         * 
+         * return capturer.
+         */
+         RSNHandshakeCapturer &get_capturer() {
+          return capturer;
+        }
```
Comment (mfontanini) [inline include/tins/crypto.h]: Spacing too xD

**Comment by mfontanini** [comment]
Sorry about the comments. I'm not trying to be mega pedantic, it's just the general syntax guidelines, like spacing, etc.

**Comment by prasad24582** [comment]
No problem Matias, its my mistake. Am sorry for making silly mistakes. I thought kept in proper spacing. Now, got to know how to make proper spacing.
Done changes as per your comments, Please check them. 
Sorry to trouble you.

**Comment by prasad24582** [comment]
Closing this, am planning to pull request because while deriving class from WPA2Decrypter getting errors. Will check and open another Pull request. Thank you.

**Pull request closed without merging**